core/store/src/trie/prefetching_trie_storage.rs

```diff
@@ -0,0 +1,417 @@
+            None => true,
+        };
+        if full {
+            return PrefetcherResult::MemoryLimitReached;
+        }
+        match guard.slots.entry(key) {
+            Entry::Occupied(entry) => match entry.get() {
+                PrefetchSlot::Done(value) => PrefetcherResult::Prefetched(value.clone()),
+                PrefetchSlot::PendingPrefetch | PrefetchSlot::PendingFetch => {
+                    PrefetcherResult::Pending
+                }
+            },
+            Entry::Vacant(entry) => {
+                entry.insert(set_if_empty);
+                guard.size_bytes += PREFETCH_RESERVED_BYTES_PER_SLOT;
+                PrefetcherResult::SlotReserved
+            }
+        }
+    }
+}
+
+impl PrefetchApi {
+    pub fn new(parent: &TrieCachingStorage) -> Self {
+        Self {
+            work_queue: Arc::new(Mutex::new(BoundedQueue::new(MAX_QUEUED_WORK_ITEMS))),
+            prefetching: parent.prefetching.clone(),
+            stop_io: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    /// Returns the trie key back if queue is full.
+    pub fn prefetch_trie_key(&self, trie_key: TrieKey) -> Option<TrieKey> {
+        self.work_queue.lock().expect(POISONED_LOCK_ERR).put(trie_key)
+    }
+
+    pub fn start_io_thread(
+        &self,
+        parent: &TrieCachingStorage,
+        trie_root: CryptoHash,
+    ) -> std::thread::JoinHandle<()> {
+        let prefetcher_storage = TriePrefetchingStorage::new(
+            parent.store.clone(),
+            parent.shard_uid,
+            parent.shard_cache.clone(),
+            self.prefetching.clone(),
+        );
+        let stop_io = self.stop_io.clone();
+        let work_queue = self.work_queue.clone();
+        std::thread::spawn(move || {
+            // `Trie` cannot be sent across threads but `TriePrefetchingStorage` can.
+            //  Therefore, construct `Trie` in the new thread.
+            let prefetcher_trie = crate::Trie::new(Box::new(prefetcher_storage), trie_root, None);
+
+            // Keep looping until signalled to stop.
+            while !stop_io.load(Ordering::Acquire) {
+                // Note: do not merge the two lines below, lock must be released in between.
+                let maybe_trie_key = work_queue.lock().expect(POISONED_LOCK_ERR).pop();
+                if let Some(trie_key) = maybe_trie_key {
+                    let storage_key = trie_key.to_vec();
+                    if let Ok(Some(_value)) = prefetcher_trie.get(&storage_key) {
+                        near_o11y::io_trace!(count: "prefetch");
+                    } else {
+                        // This may happen in rare occasions and can be ignored safely.
+                        // See comments in `TriePrefetchingStorage::retrieve_raw_bytes`.
+                        near_o11y::io_trace!(count: "prefetch_failure");
+                    }
+                } else {
+                    std::thread::sleep(Duration::from_micros(10));
```

Comment:
We expect local SSD latency to be in the order of a few times 10us. But RocksDB might have things cached in memory buffers already, so it's hard to say what number we need exactly.

I think using a crossbeam channel (instead of `ArrayQueue` that I've done for now) would be the right choice. The reason I am holing back is that I don't know how to clear all pending requests in a channel. This is required between transactions and receipts, as well as between local/delayed/incoming receipts.
I can think of some hacky ways but then I'd rather have a manual spinning loop rather than the hacks that I have in mind right now. Do you perhaps have an idea how to make `clear` work with a channel?

Reply:
I think just loop-receiving with a Receiver and dropping the received things makes sense, and would be cleaner than a busy-wait loop :)